### ISSUE TITLE
Karpenter is incompatible with FSX CSI driver fsx.csi.aws.com/agent-not-ready:NoExecute startup taint

### ISSUE THREAD

**Issue by ajayvohra2005** [aws/karpenter-provider-aws#5293]
### Description

**Observed Behavior**:

[FSx CSI Driver node startup taint](https://github.com/kubernetes-sigs/aws-fsx-csi-driver/blob/master/docs/install.md#configure-node-startup-taint) can't be used with Karpenter

There are potential race conditions on node startup (especially when a node is first joining the cluster) where pods/processes that rely on the FSx for Lustre CSI Driver can act on a node before the FSx for Lustre CSI Driver is able to start up and become fully ready. To combat this, the FSx for Lustre CSI Driver contains a feature to automatically remove a taint from the node on startup. Users can taint their nodes when they join the cluster and/or on startup, to prevent other pods from running and/or being scheduled on the node prior to the FSx for Lustre CSI Driver becoming ready.

This feature is activated by default, and one can use the taint `fsx.csi.aws.com/agent-not-ready:NoExecute` (any effect will work, but NoExecute is recommended) to use this feature.

**The use of  `fsx.csi.aws.com/agent-not-ready:NoExecute` is not possible with Karpenter**, because tainting  a NodePool with `fsx.csi.aws.com/agent-not-ready:NoExecute` prevents Karpenter from spinning up a node from the NodePool.  If one does not use the  `fsx.csi.aws.com/agent-not-ready:NoExecute` taint, then, depending on the race-condiiton outcome, one gets  `UnexpectedAdmissionError` error on the scheduled Pod. 

**Expected Behavior**:

Karpenter should ignore `fsx.csi.aws.com/agent-not-ready:NoExecute` taint while determining affinity for a NodePool, because this taint is going to be automatically removed by the FSx CSI. Perhaps, `agent-not-ready` taints should be ignored for all cases, with the assumption the agent will automatically remove them when ready.

**Reproduction Steps** (Please include YAML):

It is very hard to offer an easily reproducible test case, because it is entirely based on a race condition.

**Versions**:
- Chart Version:
- Kubernetes Version (`kubectl version`): 0.32.3

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tzneal** [comment]
Have you looked at using the startup taints on the NodePool?

```
      # Provisioned nodes will have these taints, but pods do not need to tolerate these taints to be provisioned by this
      # NodePool. These taints are expected to be temporary and some other entity (e.g. a DaemonSet) is responsible for
      # removing the taint after it has finished initializing the node.
      startupTaints:
        - key: fsx.csi.aws.com/agent-not-ready
          effect: NoExecute
```

**Comment by ajayvohra2005** [comment]
No I have not tried that. I was not aware of `startupTaints` option. Let me try that.

**Comment by ajayvohra2005** [comment]
I added the startupTaints as described above, and also tried `NoSchedule` effect, to the NodePool, as shown below:

```
...
 Startup Taints:
        Effect:  NoSchedule
        Key:     fsx.csi.aws.com/agent-not-ready
        Value:   true
 Taints:
        Effect:  NoSchedule
        Key:     nvidia.com/gpu
        Value:   true
...
```

This allowed Karpenter to launch the GPU nodes, but it did not fix the `UnexpectedAdmissionError`:

```
Events:
  Type     Reason                    Age                From                Message
  ----     ------                    ----               ----                -------
  Warning  FailedScheduling          3m20s              default-scheduler   0/2 nodes are available: 2 node(s) didn't match Pod's node affinity/selector. preemption: 0/2 nodes are available: 2 Preemption is not helpful for scheduling..
  Normal   Nominated                 3m19s              karpenter           Pod should schedule on: nodeclaim/gpu-dh2sd
  Normal   NotTriggerScaleUp         2m47s              cluster-autoscaler  pod didn't trigger scale-up: 1 node(s) didn't match Pod's node affinity/selector
  Normal   Scheduled                 2m18s              default-scheduler   Successfully assigned kubeflow/jupyter-74cdc8bdd6-jw7bn to ip-xxx-xxx-xxx-xx.us-west-2.compute.internal
  Normal   Pulling                   2m16s              kubelet             Pulling image "xxxxxxxxxxxx.dkr.ecr.us-west-2.amazonaws.com/xxxx"
  Warning  UnexpectedAdmissionError  30s                kubelet             Allocate failed due to no healthy devices present; cannot allocate unhealthy devices nvidia.com/gpu, which is unexpected
  Warning  FailedMount               30s (x2 over 30s)  kubelet             MountVolume.MountDevice failed for volume "pv-fsx" : kubernetes.io/csi: attacher.MountDevice failed to create newCsiDriverClient: driver name fsx.csi.aws.com not found in the list of registered CSI drivers
  Warning  FailedMount               30s (x2 over 30s)  kubelet             MountVolume.SetUp failed for volume "nginx-htpasswd" : object "kubeflow"/"nginx-htpasswd" not registered
  Warning  FailedMount               30s (x2 over 30s)  kubelet             MountVolume.SetUp failed for volume "nginx-conf" : object "kubeflow"/"nginx-conf" not registered
  Warning  FailedMount               30s (x2 over 30s)  kubelet             MountVolume.SetUp failed for volume "kube-api-access-glvvc" : object "kubeflow"/"kube-root-ca.crt" not registered

```

This is definitely a race condition issue, because an automatic restart resolves it, as shown below:

```
NAME                            READY   STATUS                     RESTARTS        AGE
jupyter-74cdc8bdd6-7xgtz        3/3     Running                    0               7m27s
jupyter-74cdc8bdd6-jw7bn        0/3     UnexpectedAdmissionError   0               10m
```

The real challenge is in cases where restart on failure is not allowed, because it is not feasible. In those case, this requires a manual restart.

Surprisingly, this does not happen when Cluster Auto Scaler launches the same instance type nodes, even without `startupTaints`.  I only see this with Karpenter launched nodes.

**Comment by jonathan-innis** [comment]
> This is definitely a race condition issue, because an automatic restart resolves it, as shown below

Is there a way that we can get timestamps around when the pod was scheduled relative to when the startup taint was applied to the node? I'm guessing from looking at the events that you shared, but I'm assuming that the pod is getting scheduled to the node _before_ the startup taint is actually applied to the node. So the order looks like
1. Pod goes pending due to no nodes
2. Node is launched by Karpenter with the startup taint 
3. Node joins the cluster
4. Node becomes ready 
5. Pod gets scheduled to the Node
6. Node gets tainted with the startup taint

Can you confirm if this is the ordering from referring to apiserver audit logs to see which order things were applied and evaluated in?

**Comment by ajayvohra2005** [comment]
I think the error for the sequence is as follows:

1. Karpenter applies the FSx startup taint correctly after the node is launched
2. FSx CSI Daemon set executes on the Karpenter launched Node and removes the startup taint
3. Application Pod gets scheduled and starts executing on the Karpenter launched Node
4. While the application Pod is in `ContainerCreating` state,  or maybe even in the `Running` state, the Karpenter launched node reboots (unknown reason)
5. When application Pod starts again after the Node reboots, the FSx CSI Daemon set Pod has not yet completed FSx CSI driver initialization, so the error happens, and Pod moves to `UnexpectedAdmissionError`  state.

Root cause appears to be the Node reboot.

**Comment by ajayvohra2005** [comment]
I have found the root cause and fixed it. The root cause is that  `AmazonSSMManagedInstanceCore` IAM Policy was added to the Karpenter Node role. This was causing amazon-ssm-agent to be started on the Karpeneter launched node.

It was the `amazon-ssm-agent` that was rebooting the Karpenter launched node after the startup taint had already been removed and the Pod had already been scheduled on the Node.  

After the reboot,  the sequence of startup between FSx CSI Daemon set pod and the application pod had a race condition, because startup taint is no longer in play, leading to the `UnexpectedAdmissionError`.  Removing `AmazonSSMManagedInstanceCore` IAM Policy  from the Node role fixes the issue.
